architecture/cluster-fulfillment.md

```diff
@@ -1 +1,319 @@
+Cluster that reconciles ClusterOrder resources
+(`cloudkit-controller/internal/controller/clusterorder_controller.go`).
+
+**ClusterOrder Custom Resource** (`cloudkit-controller/api/v1alpha1/clusterorder_types.go`):
+
+The ClusterOrder CRD defines:
+- `TemplateID`: Identifies which cluster template to use
+- `TemplateParameters`: JSON-encoded map of template-specific parameters
+- `NodeRequests`: Array of node request objects specifying resource class and quantity
+
+**Reconciliation Loop**:
+
+When a ClusterOrder is created or updated, the controller performs these steps:
+
+1. **Status Initialization**: Sets the phase to "Progressing" and initializes status conditions
+
+2. **Infrastructure Preparation**: Creates prerequisite Kubernetes objects:
+   - **Namespace**: A dedicated namespace for the cluster (typically named after the cluster ID)
+   - **ServiceAccount**: Identity for cluster-related automation
+   - **RoleBindings**: RBAC permissions for the service account to manage cluster resources
+
+3. **EDA Webhook Trigger**: Calls the Event Driven Ansible webhook endpoint to initiate cluster provisioning:
+   - The webhook payload includes the complete ClusterOrder specification.
+
+4. **HostedCluster Monitoring**: After the webhook is triggered, the controller watches for a HostedCluster resource (created by the Ansible automation) and monitors its conditions:
+   - Waits for the control plane to become available
+   - Monitors that the cluster is not degraded
+   - Tracks node pool readiness
+
+5. **Status Synchronization**: Continuously updates the ClusterOrder status based on the HostedCluster state, including:
+   - Phase transitions (Progressing → Ready or Failed)
+   - Condition updates
+   - Cluster reference information (namespace, resource names)
+
+6. **Deletion Handling**: When a ClusterOrder is deleted:
+   - Triggers the deletion webhook to EDA
+   - Ensures all associated cluster resources are cleaned up
+   - Uses finalizers to prevent premature deletion
+
+### Ansible Automation Platform - Cluster Provisioning
+
+Event Driven Ansible (EDA) running on AAP receives webhook events from the OSAC
+Controller and orchestrates the actual cluster provisioning.
+
+**EDA Rulebook** (`cloudkit-aap/rulebooks/cluster_fulfillment.yml`):
+
+The rulebook listens on port 5000 and defines rules for cluster lifecycle events:
+- Endpoint: `create-hosted-cluster` → Triggers the cluster creation workflow
+- Endpoint: `delete-hosted-cluster` → Triggers the cluster deletion job
+
+When triggered, EDA launches the appropriate workflow template in AAP.
+
+**Cluster Creation Playbook** (`cloudkit-aap/playbook_cloudkit_create_hosted_cluster.yml`):
+
+The main cluster creation workflow consists of these phases:
+
+1. **Retrieve Cluster Order Details**:
+   - Fetches the ClusterOrder object from Kubernetes
+   - Generates a unique lock holder ID to prevent concurrent modifications
```

Comment:
I would omit this as being an unnecessary level of detail. The important bit is in the following section, "acquires a cluster lock...".